Touch Fuctions + Tests

rkabrick opened the pull request.

rkabrick committed:
Touch Fuctions + Tests

rkabrick committed:
Added function search function for finding all uses of entities

rkabrick committed:
Updated tests which don't account for whitespace

rkabrick committed:
Updated Tests

TApplencourt commented on tests/test_global.sh:
```diff
@@ -2,11 +2,14 @@
+
+./irpc.py ./examples/touch.irp.c > tests/touch.c
+diff -w -b -B ./tests/gold.touch.c tests/touch.c
 
```
Nice change! Thanks!

TApplencourt commented:
Ok, I don't understand why ba4d28f is failing...

TApplencourt committed:
Delete touch.c

TApplencourt commented on tests/gold.touch.c:
```diff
@@ -0,0 +1,41 @@
+  a = 20;
+  touch_a();
+  b = a;
+}
```
It should have a call to `provide_a()`

TApplencourt commented on tests/gold.touch.c:
```diff
@@ -0,0 +1,41 @@
+
+  a = 20;
+  touch_a();
+  b = a;
```
It should have a call to `provide_a()`

TApplencourt commented on irpc.py:
```diff
@@ -120,18 +178,60 @@ def add_provider_call(funcdef: FuncDef,
+
+
+def find_touches(filename):
+    l_touch = set()
```
You should recurse the AST and find is we have a funccall `touch_{provider}`

TApplencourt commented on irpc.py:
```diff
@@ -120,18 +178,60 @@ def add_provider_call(funcdef: FuncDef,
+    l_touch = set()
+    with open(filename, 'r') as input:
+        for line in input:
+            if "touch_" in line:
```
Indeed, You can have false positive like
```
int my_variable_touch_a = 10
```
This is not a fun call, but will pass your test

TApplencourt commented on irpc.py:
```diff
@@ -109,6 +103,70 @@ def hoist_declaration(main: FuncDef,
+            # Recursive call followed by updating the dictionary
+            for k,v in find_instances(node, l_ent, c, j).items():
+                d[k] += v
+
```
Nice comments! Thanks!

TApplencourt commented on examples/touch.irp.c:
```diff
@@ -0,0 +1,15 @@
+
+int main(){
+  printf("b = %i", b);
+}
```
May we should do A -> B -> C where C touch A and use b?
So we will need to call `provide_b` after the `touch_a()`, and our test will be more robust (testing more things).
Or we can have two tests. More test is better :)

TApplencourt commented on tests/gold.touch.c:
```diff
@@ -0,0 +1,41 @@
+  printf("b = %i", b);
+}
+
+
```
Does this thing compile? We use `touch_a()` before declaring it.

rkabrick replied:
Yes... you had mentioned you didn't want the user to have to declare the touch functions so as of now, it works like that.

TApplencourt commented:
> I don't understand why this is failing either... The error message in the log makes no sense especially with the included flags. Also, the test passes every time I run it locally

I guess it's something to do with the randomness of `set`. But yes, it strange. I will investigate more...

TApplencourt commented on tests/test_global.sh:
```diff
@@ -2,11 +2,14 @@
+diff -w -b -B ./tests/gold.simple.c tests/simple.c
+
+./irpc.py ./examples/touch.irp.c > tests/touch.c
+diff -w -b -B ./tests/gold.touch.c tests/touch.c
```
Please add a test for the compilation of the `touch.c`

Also, You can split test_global in `test_simple` and `test_touch`. Maybe more readable

TApplencourt commented on tests/test_global.sh:
```diff
@@ -2,11 +2,14 @@
 
 # Output checking
-diff ./tests/gold.simple.out tests/simple.out
+diff -w -b -B ./tests/gold.simple.out tests/simple.out
```
Nice change! Thanks!

rkabrick committed:
Trying diff with another flag 

-I will hopefully ignore lines that are the same. This is important because the work log on Travis shows the output of the diff being the same lines

rkabrick committed:
Getting rid of -I flag in test

-I broke it more than it was before

rkabrick committed:
Not working

rkabrick committed:
Improved Provider Call Insertion

This commit implements:
- *find_instances*: Creates a dictionary based on the index within a compound to
keep track of all ID nodes (where every provided entity is used) throughout the AST
- *insert_provider_call*: inserts provider calls based on the occurences of ID nodes within the AST. It filters
out the entity list to prevent endless loops of a provider calling itself (ie.
provide_b() will never call the provider)
- *node_extract*: Extracts the compound from any type of AST node passed to it
